news/_posts/2021-11-15-form-theme-new-in-1.7.8.md

```diff
@@ -0,0 +1,109 @@
+---
+layout: post
+title:  "Introducing a new Form Theme for 1.7.8"
+subtitle: "An easy way to render Symfony forms"
+date:   2021-11-12
+authors:  [ matks ]
+icon: icon-lab
+tags: 
+ - 1.7.8.x
+ - feature
+ - development
+---
+
+In PrestaShop 1.7.8, we have carried out a [refactoring project](https://github.com/PrestaShop/PrestaShop/issues/16482) that modifies how we render forms in Symfony pages thanks to a new Form Theme. In this article, we'll see the benefits of this system and how it can be used in modules.
+
+## What is a Form Theme?
+
+In PrestaShop 1.7.8, most Symfony Back-office pages that feature forms are being powered by the PrestaShop  UI Kit Form theme.
+
+But what exactly is a Form Theme?
+
+To answer this question, let's dig a little bit into how [Symfony forms](https://symfony.com/doc/3.4/forms.html) are structured.
+
+Symfony decouples how forms are rendered from how forms are being constructed.
+
+Each form in Symfony is built using an object FormType. This object carriers the structure of the form, here's an example:
+
+```php
+class ContactType extends FormType
+{
+    
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('title', TranslatableType::class, [
+                'label' => $this->trans('Title', 'Admin.Global'),
+                'options' => [
+                    'constraints' => [
+                        new Length([
+                            'max' => $maxLength,
+                            'maxMessage' => sprintf('This field cannot be longer than %limit% characters', $maxLength)
+                        ])
+                    ]
+                ]
+            ])
+            ->add('email', EmailType::class, [
+                'label' => $this->trans('Email address', 'Admin.Global'),
+                'required' => false,
+                'constraints' => [
+                    new Email(['message' => '%s is invalid.'])
+                ],
+            ])
+        ;
+```
+
+The above php code basically states that this form contains
+- a 'title' input that has an upper limit length
+- an 'email' input that must contain a valid email
+
+Here we have the structure of our form. However this does not specify the HTML structure of the form, such as how inputs should be rendered, should they be wrapped in div, what CSS class to apply...
+
+In Symfony, you can define the HTML structure of a form either by writing it down, tag by tag, or you can delegate this responsibility to a Form Theme.
+
+The Form Theme is a set of macros that describes how each item of a form is rendered into HTML. It will state how one input, properly specified, should be rendered as an HTML structure. For example, it controls how to render in HTML a 'textarea' input with a validation message and a label.
```

Comment:
```suggestion
The Form Theme is a set of macros that describes how each form item is rendered into HTML. It will state how one form field, properly specified, should be rendered as an HTML structure. For example, it controls how to render in HTML a 'textarea' field with a validation message and a label.
```